### ISSUE TITLE
Leak of durable objects through old virtual and heap references after upgrade

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#9338]
## What is the Problem Being Solved?

A durable object can be reachable by being directly exported, or by a local reference from a durable object, virtual-only object or heap object. 

During an upgrade, all virtual-only and heap references are severed. While we track export vs heap vs "virtual/durable" (the 3 "pillars"), we do not currently differentiate virtual-only from durable in the latter. We also do not have a way to enumerate the objects which only have the virtual-only or heap pillar, that was severed by the upgrade. As such these "previous incarnation only" references cause the accumulation of garbage when upgrades occur.

The problem described here is related to information maintained by the vat, and unknown to the kernel. It's different from issues like #7212 which describe cases where the kernel realizes some exported references are no longer reachable after an upgrade.

Even before the removal of `stopVat` (#6650 closed by #7244) we've been unable to collect some of these cases because we lacked the information necessary to find these "incarnation only references".

### Track virtual-only
Currently each virtual/durable object maintains a `vom.rc.${baseRef}` entry in the vatStore which contains the total ref count from other vrefs. This forms the "virtual pillar". We need to extend this to differentiate the virtual-only from durable case. We also need to understand if the virtual-only count is from the current or a previous incarnation.

We can do this in a backwards compatible manner by transforming the field `vom.rc.` into the following: `${totalCount},${virtualOnlyCount},${virtualOnlyIncarnationNum}`. When touching the field, the VOM can find out that the virtual-only count comes from an old incarnation, and zero it after substracting it from the total count. For a virtual-only objects, the `totalCount` and the `virtualOnlyCount` would always be the same.

### Track no "upgradable reference" object set

Tracking virtual-only and heap references does not help identifying which virtual/durable objects are no longer referenced after an upgrade. For that we could introduce a durable set recording the vrefs which do not have a durable or export pillar, let's call it "incarnation only references set". Any time an object drops either its export pillar (aka simplified `get(vom.es.${baseRef}) !== 'r'`) or its durable pillar (`totalCount - virtualOnlyCount === 0`), we would check the other pillar, and if both are gone, the vref of the object would be added to the `incarnationOnlyRefSet`. Exporting the object or referencing it from a durable object would remove the object from the set.

When the `totalCount` drops to zero, the object can be added to the `possiblyDeadSet` as only the heap pillar would possibly remain. Effectively this `incarnationOnlyRefSet` set becomes a precursor state to being in the `possiblyDeadSet`, and the `isVirtualObjectReachable` could be updated to check this single "incarnation only set".

To support slow deletion (#8417), this "incarnation only set" could in fact be a map which records if there are remaining virtual-only references `totalCount !== 0`, and possibly in which incarnation these virtual-only references existed. Then the slow deletion logic could add the vrefs to `possiblyDeadSet` in stages. It would also avoid a leak through the `possiblyDeadSet` which is heap only.

### Tracking virtual-only definitions

This part is a little less fleshed out, but durably tracking virtual/durable instances is not sufficient to clean up all garbage. The definitions of virtual-only kinds create a `vom.vkind` entry each time. Once all objects of that vkind have been deleted from the vatStore, we should be safe to delete this entry as well. I'm not sure how to track this, or if it's worth the trouble.

## Security Considerations

None I can think of right now

## Scaling Considerations

Adding a large amount of objects to this "incarnation only set" still causes a lot of syscalls (at least while vatStore operations require syscalls). For the case of collection deletion, the "Rate-Limited Collection Deletion" section of #8417 describes a list of collections which need to be slow deleted, and we could potentially merge it with the "incarnation only set" (e.g. when encountering a non-empty collection in the set, remove N entries and place them in the set).

## Test Plan

TBD

## Upgrade Considerations

TLDR: liveslots must leave enough information behind so that successors can clean up incarnation specific data.

One thing this approach doesn't help as described is how to collect virtual-only objects (or heap object) which were exported (see #6696). However we could maintain a similiar durable list of non-durable vrefs which were exported, and in the new incarnation, we would slowly iterate entries from the previous incarnations which have a `totalCount === 0`, drop the export, and add them to the "incarnation only set" (the non 0 totalCount would be dropped when their virtual references get dropped). #7170 implemented this recognition kernel-side after an upgrade, but this might cause a large amount of gc actions at once. A kernel side mitigation to large gc actions is described in #8930, but having vats being good citizens in the first place would help too (and avoid some of the layering violations).

**Comment by mhofman** [comment]
Might be a similar problem for recognizable references that were kept in ephemeral virtualized weak collections previous to an upgrade

**Comment by warner** [comment]
I wanted to capture an idea from today's meeting. The topic was how to clean up after a previous incarnation, where some large number of durable objects were only kept alive by ephemeral or merely-virtual references, all of which are now gone.

The straightforward but really expensive approach is for the new incarnation to do a mark-and-sweep of every durable object, to find the ones that have no references. The new incarnation does not see the old incarnation's virtuals or ephemerals, so the lack of those references will allow the durables to be collected. But that's multiple DB queries for every single durable object.

We've talked about having an offline process do the expensive mark-and-sweep work to identify unreachable objects, and then submitting a list of alleged vrefs to the chain, whereupon the kernel (or liveslots) could verify that each one is actually unreferenced and delete it. The API would be something like `controller.scanAllegedDeadVrefs(vatID, vrefs..)`. That's a lot cheaper for the vat- side code, it's only doing DB queries for the allegedly-dead vrefs. But the list of vrefs might be quite large (unless the caller deliberately partitions their real list into disjoint sets, and does the cleanup over multiple calls, spread out over time).

I've wanted to replace the (possibly long) list of vrefs with a bloom filter. (Ok, really I just desperately want to use a bloom filter in production some day before I retire, but every place I've tried so far required the opposite false-negative polarity, curses!). The expensive way to do this would be `.scan(vatID, vrefFilter)`, where the recipient already has a list of all vrefs, but must do a DB scan to build up that list. This would make the API arguments small, but costs the vat as many DB references as the full mark-and-sweep.

The idea I had today was to make the arguments be `(vatID, kindID, firstInstance, lastInstance, vrefFilter)`. So we partition the objects by Kind, and then we can build the list of potential vrefs by cycling a counter instead of doing a DB scan for `vs.vom.${vref}` state-record keys, and subjecting each potential vref to the bloom filter before doing the refcount checks. Maybe even make it a list of `{ kindID, first, last }` tuples, with a single `vrefFilter` applied to all the vrefs thus generated. We could afford to generate thousands of vrefs this way without making the txn args very large, and we'd only do DB work for the subset that passed the filter. (We should probably also include a `numVrefs` count, and have the loop stop after checking that many, and have the implementation refuse to consider more than a few thousand, as a protective measure).

**Comment by mhofman** [comment]
> We've talked about having an offline process do the expensive mark-and-sweep work to identify unreachable objects, and then submitting a list of alleged vrefs to the chain, whereupon the kernel (or liveslots) could verify that each one is actually unreferenced and delete it.

Thinking more about this, that wouldn't work. How would the kernel or liveslots cheaply verify there are no durable references. Afaict, the only option is tracing the durable roots?

**Comment by warner** [comment]
Yeah, you're right, it would only work in the case where the refcounts are cheap to access and accurate/complete, and would only be useful in the case where the vrefs that might have zero refcounts were expensive to determine.
